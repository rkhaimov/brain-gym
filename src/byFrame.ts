import { Observable, OperatorFunction } from 'rxjs';

export function byFrame<TValue>(): OperatorFunction<TValue, TValue> {
  return (observable) =>
    new Observable((emitter) => {
      let id: undefined | ReturnType<typeof requestAnimationFrame>;
      const subscription = observable.subscribe((value) => {
        if (id) {
          cancelAnimationFrame(id);

          id = undefined;
        }

        id = requestAnimationFrame(() => emitter.next(value));
      });

      return () => {
        subscription.unsubscribe();

        if (id) {
          cancelAnimationFrame(id);
        }
      };
    });
}