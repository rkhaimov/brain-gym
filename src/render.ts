import { applyChangesToRealElement } from './applyChangesToRealElement';
import { VirtualElement } from './elements';

export function createRenderer() {
  let prev: VirtualElement | undefined;

  return (curr: VirtualElement, parent: HTMLElement) => {
    applyChangesToRealElement(prev, curr, parent);

    prev = curr;

    return parent;
  };
}
