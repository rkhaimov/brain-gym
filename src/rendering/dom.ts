import { Attributes, Listeners, Props, RenderingTree } from './types';

export function commit(container: HTMLElement, tree: RenderingTree) {
  if (typeof tree === 'string') {
    container.innerHTML = tree;

    return;
  }

  const node = document.createElement(tree.tag);

  for (const [attr, value] of Object.entries(tree.attributes)) {
    node.setAttribute(attr, value);
  }

  for (const [event, handler] of Object.entries(tree.listeners)) {
    node.addEventListener(event, handler);
  }

  tree.children.forEach((child) => commit(node, child));

  container.appendChild(node);
}

export function propsToAttributes(props: Props): Attributes {
  if (props === null) {
    return {};
  }

  const attributes = Object.entries(props).filter(
    (pair): pair is [string, string] => typeof pair[1] !== 'function'
  );

  return Object.fromEntries(attributes);
}

export function propsToListeners(props: Props): Listeners {
  if (props === null) {
    return {};
  }

  const listeners = Object.entries(props)
    .filter(
      (pair): pair is [string, () => void] => typeof pair[1] === 'function'
    )
    .map(([key, value]) => [key.slice(2).toLowerCase(), value]);

  return Object.fromEntries(listeners);
}
