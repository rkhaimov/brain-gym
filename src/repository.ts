export async function searchOptions(search: string) {
  console.log('Start Searching with', search);

  await wait(Math.random() * 10_000);

  console.log('Done Searching with', search);

  return ['Option 1', 'Option 2', 'Option 3', search];
}

export function wait(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
