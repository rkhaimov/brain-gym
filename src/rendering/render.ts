import { Observable } from 'rxjs';
import { applyChangesToRealElement } from './applyChangesToRealElement';
import { byFrame } from './byFrame';
import { VirtualNodeOrString } from './createElement';

export function createRenderer() {
  let prev: VirtualNodeOrString | undefined;

  return (curr: VirtualNodeOrString, parent: HTMLElement) => {
    applyChangesToRealElement(prev, curr, parent);

    prev = curr;

    return parent;
  };
}

export function renderUItoDOM(ui$: Observable<VirtualNodeOrString>) {
  const render = createRenderer();
  const main = document.createElement('main');
  document.body.appendChild(main);

  ui$.pipe(byFrame()).subscribe((dom) => render(dom, main));
}