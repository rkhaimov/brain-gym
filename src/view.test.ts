import { fireEvent, getByRole, prettyDOM } from '@testing-library/dom';
import { ARIARole } from 'aria-query';
import { times } from 'lodash-es';
import { BehaviorSubject, map, Observable, of, switchMap } from 'rxjs';
import { byFrame } from './byFrame';
import {
  createDiv,
  createH1,
  createH2,
  createInput,
  createSpan,
  VirtualElement,
} from './elements';
import { createRenderer } from './render';
import { wait } from './repository';

test('It allows to sync h1 title and input value', async () => {
  const input$ = new BehaviorSubject('');
  const ui$ = createDiv({
    children: [
      createH1({ children: [input$] }),
      createInput({ onInput: input$, children: [] }),
    ],
  });
  const rendered = renderAndInspect(ui$);
  await wait(100);

  // Act
  const input = getByRole(
    rendered.dom,
    'textbox' as ARIARole
  ) as HTMLInputElement;

  fireEvent.input(input, { target: { value: 'Hello world' } });
  await wait(100);

  // Assert
  expect(rendered.output).toMatchSnapshot();
});

test('It allows to change element type', async () => {
  // Arrange
  const input$ = new BehaviorSubject('');

  const ui$ = createDiv({
    children: [
      input$.pipe(
        switchMap((input) =>
          input.includes('H1')
            ? createH1({ children: [of(input)] })
            : createH2({ children: [of(input)] })
        )
      ),
    ],
  });

  const rendered = renderAndInspect(ui$);
  await wait(100);

  // Act
  input$.next('H1');
  await wait(100);

  input$.next('H2');
  await wait(100);

  // Assert
  expect(rendered.output).toMatchSnapshot();
});

test('It allows to change element attribute', async () => {
  // Arrange
  const input$ = new BehaviorSubject('');

  const ui$ = createDiv({
    children: [
      createSpan({ children: [input$], className: input$ }),
      createInput({ onInput: input$, children: [] }),
    ],
  });

  const rendered = renderAndInspect(ui$);
  await wait(100);

  // Act
  const input = getByRole(
    rendered.dom,
    'textbox' as ARIARole
  ) as HTMLInputElement;

  fireEvent.input(input, { target: { value: 'Hello' } });
  await wait(100);

  fireEvent.input(input, { target: { value: 'World' } });
  await wait(100);

  expect(rendered.output).toMatchSnapshot();
});

test('It can remove or add child', async () => {
  const input$ = new BehaviorSubject('0');

  const ui$ = createDiv({
    children: [
      input$.pipe(
        map((value) => parseInt(value)),
        switchMap((count) =>
          createDiv({
            children: times(count, (index) =>
              createH1({ children: [of(`${index}`)] })
            ),
          })
        )
      ),
    ],
  });

  const rendered = renderAndInspect(ui$);
  await wait(100);

  input$.next('2');
  await wait(100);

  input$.next('1');
  await wait(100);

  expect(rendered.output).toMatchSnapshot();
});

test('It can update contents of specific child', async () => {
  const input$ = new BehaviorSubject('off');

  const ui$ = createDiv({
    children: [
      input$.pipe(
        switchMap((input) => {
          if (input === 'on') {
            return createDiv({
              children: [
                createH1({ children: [of('0')] }),
                createH1({
                  children: [
                    createDiv({
                      children: [of('1')],
                    }),
                  ],
                }),
              ],
            });
          }

          return createDiv({
            children: [
              createH1({ children: [of('0')] }),
              createH1({
                children: [of('1')],
              }),
            ],
          });
        })
      ),
    ],
  });

  const rendered = renderAndInspect(ui$);
  await wait(100);

  input$.next('on');
  await wait(100);

  expect(rendered.output).toMatchSnapshot();
});

test('It can update type of specific child', async () => {
  const input$ = new BehaviorSubject('off');

  const ui$ = createDiv({
    children: [
      input$.pipe(
        switchMap((input) => {
          if (input === 'on') {
            return createDiv({
              children: [
                createDiv({
                  children: [of('0')],
                }),
                createH1({ children: [of('1')] }),
                createDiv({
                  children: [of('2')],
                }),
              ],
            });
          }

          return createDiv({
            children: [
              createH1({ children: [of('0')] }),
              createH1({ children: [of('1')] }),
              createH1({
                children: [of('2')],
              }),
            ],
          });
        })
      ),
    ],
  });

  const rendered = renderAndInspect(ui$);
  await wait(100);

  input$.next('on');
  await wait(100);

  expect(rendered.output).toMatchSnapshot();
});

test('It can update attributes of specific child', async () => {
  const input$ = new BehaviorSubject('off');

  const ui$ = createDiv({
    children: [
      input$.pipe(
        switchMap((input) => {
          if (input === 'on') {
            return createDiv({
              children: [
                createSpan({
                  className: of('0-on'),
                  children: [of('0-on')],
                }),
                createSpan({ className: of('1'), children: [of('1')] }),
                createSpan({
                  className: of('2-on'),
                  children: [of('2-on')],
                }),
              ],
            });
          }

          return createDiv({
            children: [
              createSpan({
                className: of('0-off'),
                children: [of('0-off')],
              }),
              createSpan({ className: of('1'), children: [of('1')] }),
              createSpan({
                className: of('2-off'),
                children: [of('2-off')],
              }),
            ],
          });
        })
      ),
    ],
  });

  const rendered = renderAndInspect(ui$);
  await wait(100);

  input$.next('on');
  await wait(100);

  expect(rendered.output).toMatchSnapshot();
});

function renderAndInspect(element: Observable<VirtualElement>) {
  let outputs: string[] = [];
  const dom = document.createElement('main');
  const render = createRenderer();

  const subscription = element.pipe(byFrame()).subscribe((ui) => {
    try {
      outputs.push(prettyDOM(render(ui, dom)) as string);
    } catch (e: unknown) {
      outputs = [(e as Error).message];

      subscription.unsubscribe();
    }
  });

  return {
    get output() {
      return outputs.join('\n\n');
    },
    dom,
  };
}
