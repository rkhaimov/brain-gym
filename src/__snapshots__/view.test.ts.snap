// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It allows to change element attribute 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<span[39m
      [33mclass[39m=[32m\\"\\"[39m
    [36m/>[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<span[39m
      [33mclass[39m=[32m\\"Hello\\"[39m
    [36m>[39m
      [0mHello[0m
    [36m</span>[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<span[39m
      [33mclass[39m=[32m\\"World\\"[39m
    [36m>[39m
      [0mWorld[0m
    [36m</span>[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m"
`;

exports[`It allows to change element type 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<h2 />[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<h1>[39m
      [0mH1[0m
    [36m</h1>[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<h2>[39m
      [0mH2[0m
    [36m</h2>[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m"
`;

exports[`It allows to sync h1 title and input value 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<h1 />[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<h1>[39m
      [0mHello world[0m
    [36m</h1>[39m
    [36m<input />[39m
  [36m</div>[39m
[36m</main>[39m"
`;

exports[`It can remove or add child 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<div />[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<h1>[39m
        [0m0[0m
      [36m</h1>[39m
      [36m<h1>[39m
        [0m1[0m
      [36m</h1>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<h1>[39m
        [0m0[0m
      [36m</h1>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m"
`;

exports[`It can update attributes of specific child 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"0-off\\"[39m
      [36m>[39m
        [0m0-off[0m
      [36m</span>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"1\\"[39m
      [36m>[39m
        [0m1[0m
      [36m</span>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"2-off\\"[39m
      [36m>[39m
        [0m2-off[0m
      [36m</span>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"0-on\\"[39m
      [36m>[39m
        [0m0-on[0m
      [36m</span>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"1\\"[39m
      [36m>[39m
        [0m1[0m
      [36m</span>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"2-on\\"[39m
      [36m>[39m
        [0m2-on[0m
      [36m</span>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m"
`;

exports[`It can update contents of specific child 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<h1>[39m
        [0m0[0m
      [36m</h1>[39m
      [36m<h1>[39m
        [0m1[0m
      [36m</h1>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<h1>[39m
        [0m0[0m
      [36m</h1>[39m
      [36m<h1>[39m
        [36m<div>[39m
          [0m1[0m
        [36m</div>[39m
      [36m</h1>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m"
`;

exports[`It can update type of specific child 1`] = `
"[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<h1>[39m
        [0m0[0m
      [36m</h1>[39m
      [36m<h1>[39m
        [0m1[0m
      [36m</h1>[39m
      [36m<h1>[39m
        [0m2[0m
      [36m</h1>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m

[36m<main>[39m
  [36m<div>[39m
    [36m<div>[39m
      [36m<div>[39m
        [0m0[0m
      [36m</div>[39m
      [36m<h1>[39m
        [0m1[0m
      [36m</h1>[39m
      [36m<div>[39m
        [0m2[0m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</main>[39m"
`;
